### PR TITLE
Fix some of the rustfmt fallout in Miri

### PR DESCRIPTION
**Pull request by RalfJung**
I tried to coerce rustfmt into reasonable formatting of matches, but that didn't always work...

r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/interpret/cast.rs [hunk 1]
```diff
@@ -118,17 +118,15 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
-                    src.layout.ty.is_bool()
-                        || src.layout.ty.is_char()
-                        || src.layout.ty.is_enum()
-                        || src.layout.ty.is_integral()
-                        || src.layout.ty.is_any_ptr(),
-                    "Unexpected cast from type {:?}",
-                    src.layout.ty
-                )
-            }
+            _ => assert!(
```
Comment (RalfJung) [inline src/librustc_mir/interpret/cast.rs]: It is really disappointing that I cannot stop rustfmt from doing silly things like this. :( Cc https://github.com/rust-lang/rustfmt/issues/3995

File: src/librustc_mir/interpret/memory.rs [hunk 2]
```diff
@@ -580,10 +580,9 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> Memory<'mir, 'tcx, M> {
                 let layout = self.tcx.layout_of(ParamEnv::empty().and(ty)).unwrap();
                 Ok((layout.size, layout.align.abi))
             }
-            Some(GlobalAlloc::Memory(alloc)) =>
-            // Need to duplicate the logic here, because the global allocations have
-            // different associated types than the interpreter-local ones.
-            {
+            Some(GlobalAlloc::Memory(alloc)) => {
+                // Need to duplicate the logic here, because the global allocations have
```
Comment (oli-obk) [inline src/librustc_mir/interpret/memory.rs]: I usually move such comments before the pattern if that will allow the pattern and the arm body to be on one line

File: src/librustc/mir/interpret/error.rs [hunk 3]
```diff
@@ -438,139 +438,165 @@ pub enum UnsupportedOpInfo<'tcx> {
 impl fmt::Debug for UnsupportedOpInfo<'tcx> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         use UnsupportedOpInfo::*;
+        #[rustfmt::skip] // rustfmt and long matches do not go well together
```
Comment (Centril) [inline src/librustc/mir/interpret/error.rs]: Imo the diff doesn't seem like that much of an improvement to justify this... I'm concerned about people starting to use this in various places throughout the compiler to fit their personal tastes rather than having a uniform style.

Comment (nikomatsakis) [reply src/librustc/mir/interpret/error.rs]: I tend to agree with @Centril here that I would like to avoid `#[rustfmt::skip]`. I'd rather find ways to structure the code that make rustfmt happy, or else open issues on rustfmt for extreme cases.

Comment (nikomatsakis) [reply src/librustc/mir/interpret/error.rs]: (Tweaking rustfmt.toml seems ok too, though I'd prefer to avoid even that when we can.)

File: src/librustc_mir/interpret/memory.rs (hunk 2 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/memory.rs]: The comment belongs to the body of the arm though, not the pattern.

File: src/librustc/mir/interpret/error.rs (hunk 3 above)
Comment (RalfJung) [reply src/librustc/mir/interpret/error.rs]: Uh... for me this diff is the difference between "unreadable garbage" and "reasonable code". Like, without this PR, the `write!` is formatted in one of *three* different ways depending on the length of the message and the phase of the moon. Sometimes it's entirely in the same line as the pattern, sometimes it is in a line on its own, and sometimes the `write!` is on the line of the pattern but its arguments are in separate lines. How is that *not* horrible for anyone? It's inconsistent, it's hard to parse visually. The entire point of rustfmt, I thought, is to get consistent formatting, but clearly it is doing the opposite here (and for `match`es in general).

I opened a rustfmt issue at https://github.com/rust-lang/rustfmt/issues/3995. But until that gets fixed, I don't see any good reason to let a tool's deficiency (rustfmt overall works fairly poorly on `match` arms) dictate the format.

File: src/librustc_mir/interpret/memory.rs (hunk 2 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/memory.rs]: Sure, but there can be no confusion here about what is meant

File: src/librustc_mir/interpret/place.rs [hunk 4]
```diff
@@ -391,12 +391,11 @@ where
...
             let align = match self.size_and_align_of(base.meta, field_layout)? {
                 Some((_, align)) => align,
-                None if offset == Size::ZERO =>
-                // An extern type at offset 0, we fall back to its static alignment.
-                // FIXME: Once we have made decisions for how to handle size and alignment
-                // of `extern type`, this should be adapted.  It is just a temporary hack
-                // to get some code to work that probably ought to work.
-                {
+                None if offset == Size::ZERO => {
+                    // An extern type at offset 0, we fall back to its static alignment.
```
Comment (oli-obk) [inline src/librustc_mir/interpret/place.rs]: And here the comment actually refers to both the pattern and the arm

File: src/librustc_mir/interpret/validity.rs [hunk 5]
```diff
@@ -114,12 +114,11 @@ fn write_path(out: &mut String, path: &Vec<PathElem>) {
...
             TupleElem(idx) => write!(out, ".{}", idx),
             ArrayElem(idx) => write!(out, "[{}]", idx),
-            Deref =>
-            // This does not match Rust syntax, but it is more readable for long paths -- and
-            // some of the other items here also are not Rust syntax.  Actually we can't
-            // even use the usual syntax because we are just showing the projections,
-            // not the root.
-            {
+            Deref => {
+                // This does not match Rust syntax, but it is more readable for long paths -- and
```
Comment (oli-obk) [inline src/librustc_mir/interpret/validity.rs]: Here the separation of pattern and body makes it harder to process and formats a single arm differently from all the above arms. It feels more consistent to have them all be a single line

File: src/librustc/mir/interpret/error.rs (hunk 3 above)
Comment (RalfJung) [reply src/librustc/mir/interpret/error.rs]: Also Cc @Mark-Simulacrum, whose suggestion it was to use `rustfmt::skip` on this file.

File: src/librustc_mir/interpret/place.rs (hunk 4 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/place.rs]: Kinda... though then the wording would be different: "for extern types at offset 0, we fall back". The current wording first states a fact that was established by the environment ("extern type at offset 0"), and then the conclusion, and this wording makes sense inside the arm, IMO.

File: src/librustc_mir/interpret/memory.rs (hunk 2 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/memory.rs]: But here the code should not be on the same line as the pattern anyway, as the other arms of the same match are much longer. Making them both on the same line makes it way too easy to miss the fact that there even is a branch here.

IMO, `pat => expr` should only be used if i can be used consistently for all arms of this `match`. Otherwise, getting consistent visual cluse about where each arm starts and ends is more important than saving vertical space.

File: src/librustc_mir/interpret/validity.rs (hunk 5 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/validity.rs]: Fair. Note that I didn't make this multi-line, rustfmt did. I did the minimal thing because I feared anything else might just be un-done by rustfmt again (as it did with most of my edits, see the second commit in this PR).

File: src/librustc_mir/interpret/memory.rs (hunk 2 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/memory.rs]: The next arm is on a single line, too. I don't have a strong opinion though, so marking as resolved

File: src/librustc/mir/interpret/error.rs (hunk 3 above)
Comment (pnkfelix) [reply src/librustc/mir/interpret/error.rs]: You may not like this option, but you could also rewrite the code to use this pattern:

let val;
match input {
    pat_1 => {
        val = expr_1;
    }
    pat_2 => {
        val = expr_2;
    }
    ...
}
```

At least, my informal experiments during the compiler team meeting indicated that using a statement would force `rustfmt` to keep the braces, and once you have the braces, it tended to keep the newlines that would make the `write!` inovcations match up.

Comment (RalfJung) [reply src/librustc/mir/interpret/error.rs]: So in order to please rustfmt, we'll change code to be less idiomatic (statements instead of expressions)? That seems... odd? rustfmt isn't some God-given axiom that we have to blindly follow, is it?

**Comment by RalfJung** [comment]
Another `match` that got badly hit is [this](https://github.com/rust-lang/rust/blob/ee84c30aee06a004b8d8f8d24000351e9d1cb4bf/src/librustc_mir/interpret/intrinsics.rs#L108): there are multiple match arms that have patterns like `A | B | C`, but some of them get formatted with one pattern per line, while others get formatted with all patterns on the same line. Visually scanning for a particular arm becomes impossible.

File: src/librustc_mir/interpret/memory.rs (hunk 2 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/memory.rs]: Yeah I'd like that to be in its own line as well but it's impossible. :(

File: src/librustc/mir/interpret/error.rs (hunk 3 above)
Comment (Centril) [reply src/librustc/mir/interpret/error.rs]: Fwiw, I think that rewriting beautiful functional-looking code into weird unidiomatic imperative code would be taking things too far. Maybe it's not optimal, but I don't think the style that is being replaced in this diff is *that* egregious that we need to use `val = expr_1;`.

> rustfmt isn't some God-given axiom that we have to blindly follow, is it?

No, it's just useful to be able to format-on-save anywhere, not caring about style and have it fixed for you, as well as a uniform reading experience, and reducing pure formatting changes in `git blame` over time.

Comment (RalfJung) [reply src/librustc/mir/interpret/error.rs]: >  it's just useful to be able to format-on-save anywhere

I appreciate that's useful for folks who went through the hoops of setting that up. Even as someone who didn't, I appreciate the consistent formatting of function signatures, the sorted imports, stuff like that.

But when considering `match`es like this, I honestly don't understand how you can call this a "uniform reading experience" -- there's nothing uniform about how rustfmt formats this match. As far as I am concerned, the formatting of this match after rustfmt is literally worse than anything I ever saw anywhere in the rustc codebase before rustfmt. The way it formats matches *will* make it much harder for me to work on and review this code-base, due to how it degrades readability. I appreciate that is subjective, but the part where it is very inconsistent and using different styles to do the exact same thing is objectively neither uniform nor consistent.

I think the style in this diff is by far egregious enough to justify `rustfmt::skip`. I also don't get the principled opposition to that: we don't usually make ourselves slaves to our own tools, do we? Tidy has had exceptions that were and are applied where reasonable. Not allowing reasonable exceptions would be quite unprecedented. In fact, I count around 30 occurrences of `rustfmt::skip` in the rustc repo already (excluding tools and tests). Many of them are for some sort of table or another (but in some sense, this `match` *is* a table), but [this one is not](https://github.com/rust-lang/rust/blob/f48e576756cd0c360e5522974fc8d5867b439092/src/liballoc/boxed.rs#L595). And that makes sense -- rustfmt is a great tool, but like all tools it has its limitations. We should apply the tool where it makes sense, but not mindlessly submit ourselves to whatever the tool happens to do.

**Comment by RalfJung** [comment]
I guess ultimately this is up to the compiler team, so maybe we should nominate? The higher-level point here is to come up with some kind of a policy for when `rustfmt::skip` is acceptable. Also see the discussion at https://github.com/rust-lang/rust/pull/67833#discussion_r362851169.

I just picked that one match as I think it is the most obvious and worst case of match-formatting gone wrong. But also see [this commit](https://github.com/rust-lang/rust/pull/67833/commits/ff5b102e008dde17a8ec0ed0c364ac647d835a3e) of rustfmt edits that all decrease `match` consistency (so if this was my codebase I'd not accept a PR landing such code), and [this commit](https://github.com/RalfJung/rust/commit/e9e20900998448b7b1706a8bfe9853baee9f593d) manually formatting two more `match`es.

**Comment by Mark-Simulacrum** [comment]
I would personally be in favor of landing the skips as a stopgap and seeing what we can do to upstream fixes into rustfmt, even if those are gated behind some option.

In particular, I find that I much prefer to enable people to contribute and we should have a policy that skip can be applied anywhere but should be tagged with a rustfmt issue associated with it. If it becomes more widespread, we can move to trying to limit the cases, but my hope is that it touches 1% of the code and we can mostly not care.

**Review by oli-obk — APPROVED** [review]

File: src/librustc/mir/interpret/error.rs (hunk 3 above)
Comment (pnkfelix) [reply src/librustc/mir/interpret/error.rs]: I'll just note that sometimes I find `let val; ...; val = expr` *more* readable.

(Not when its mutable state; that's almost never more readable. I'm talking about lazily initialized state here.)

Why do I find it more readable? Well, sometimes the `let val = { ... }; ends up with something where the result expression is so far from the binding that it doesn't fit in my editor's view at once, which means I've effectively lost the relationship (or at least have to keep in in my mental cache).

Comment (Centril) [reply src/librustc/mir/interpret/error.rs]: > Why do I find it more readable? Well, sometimes the `let val = { ... }; ends up with something where the result expression is so far from the binding that it doesn't fit in my editor's view at once, which means I've effectively lost the relationship (or at least have to keep in in my mental cache).

Split your functions Felix, so that "doesn't fit" doesn't happen :P

**Comment by pnkfelix** [comment]
Hey @RalfJung , @nikomatsakis and I were musing that if `rustfmt` had a config option that told it to not rewrite the brace-styling on match arms (i.e. don't replace `PAT => { EXPR }` with `PAT => EXPR,`) then that might address the resulting formatting issues you have here, as long as you were willing to write your own code here to use braces. (But you would not, I believe, need to add assignment statements to your code in the manner I outlined earlier.)

Would that be a reasonable way to address this, from your point of view, if it does work?

**Comment by RalfJung** [comment]
It would certainly be a big improvement, yes. :)

I was hoping to also lobby for `match_arm_blocks = false` to make rustfmt less aggressive about adding braces, not sure how these options would end up behaving together.

**Comment by pnkfelix** [comment]
@RalfJung oh, right; my intent was that the config flag would apply in *both* directions. I.e., if the flag is set, then `rustfmt` should err on the side of leaving the brace-style as is. So if they're present, leave them there, and if they're absent, don't insert them.

**Comment by oli-obk** [comment]
[Discussed at the compiler team meeting](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/weekly.20meeting.202020-01-16.20.2354818/near/185827605). There isn't much opposition to `#[rustfmt::skip]` addings, but also noone really in favour. So we won't be accepting this PR to prevent churn. Changes to rustfmt.toml causing widespread changes would still be ok.

**Comment by RalfJung** [comment]
Well, that is disappointing, but I guess there is little I can do here. :(

I'll reopen the PR without the `rustfmt::skip` part -- and I will try to never open that `error.rs` file ever again...

File: src/librustc/mir/interpret/error.rs (hunk 3 above)
Comment (pnkfelix) [reply src/librustc/mir/interpret/error.rs]: > Split your functions Felix, so that "doesn't fit" doesn't happen :P

Sure, when that's an option, it can work well. And I suppose you might also suggest that I ensure that the enums I process never have more than four or five variants.

but sometimes you have to play the hand you are dealt, and I'm glad that I have `let x; ... x = ...;` as a tool in my back pocket, even you all aren't willing to use it yourself.

Comment (Centril) [reply src/librustc/mir/interpret/error.rs]: > And I suppose you might also suggest that I ensure that the enums I process never have more than four or five variants.

Not really, but you can split `match` arms into separate functions when the `match` gets too long, e.g. as in https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_typeck/check/expr.rs.html#213-289 or https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_typeck/check/pat.rs.html#119-219 which were unreadable messes before I split those functions apart.
